### PR TITLE
Add support queue in upstreams for plus

### PR DESCRIPTION
**Pull request by B-V-R**
### Proposed changes
NGINX Plus supports specifying a queue for an upstream. If an upstream server cannot be selected immediately while processing a request, the request will be placed into the queue. See http://nginx.org/en/docs/http/ngx_http_upstream_module.html#queue

### Checklist
Before creating a PR, run through this checklist and mark each as complete.

- [x] I have read the [CONTRIBUTING](https://github.com/nginxinc/kubernetes-ingress/blob/master/CONTRIBUTING.md) doc
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked that all unit tests pass after adding my changes
- [x] I have updated necessary documentation
- [x] I have rebased my branch onto master
- [x] I will ensure my PR is targeting the master branch and pulling from my branch from my own fork

### CODE REVIEW

File: pkg/apis/configuration/validation/validation_test.go [hunk 1]
```diff
@@ -2011,3 +2011,74 @@ func TestRejectPlusResourcesInOSS(t *testing.T) {
...
+	for _, test := range tests {
+		allErrs := validateQueue(test.upstreamQueue, test.fieldPath, test.isPlus)
+		if len(allErrs) != 0 {
+			t.Errorf("validateQueue() returned errors %v for valid input for the case of %s", allErrs, test.msg)
+		}
+	}
+}
+
+func TestValidateQueueFails(t *testing.T) {
+
```
Comment (Dean-Coakley) [inline pkg/apis/configuration/validation/validation_test.go]:
```suggestion
```

File: docs/virtualserver-and-virtualserverroute.md [hunk 2]
```diff
@@ -243,6 +245,24 @@ See the [proxy_buffers](https://nginx.org/en/docs/http/ngx_http_proxy_module.htm
 | ----- | ----------- | ---- | -------- |
 | `enable` | Enables HTTPS for requests to upstream servers. The default is `False`, meaning that HTTP will be used. | `boolean` | No |
 
+### Upstream.Queue
+
+The queue field configures a queue. A client request will be placed into the queue if an upstream server cannot be selected immediately while processing the request.
```
Comment (Rulox) [inline docs/virtualserver-and-virtualserverroute.md]:
```suggestion
The queue field configures a queue. A client request will be placed into the queue if an upstream server cannot be selected immediately while processing the request:
```

File: docs/virtualserver-and-virtualserverroute.md [hunk 3]
```diff
@@ -243,6 +245,24 @@ See the [proxy_buffers](https://nginx.org/en/docs/http/ngx_http_proxy_module.htm
...
+### Upstream.Queue
+
+The queue field configures a queue. A client request will be placed into the queue if an upstream server cannot be selected immediately while processing the request. 
+
+```yaml
+size: 10
+timeout: 60s
+```
+
+See [`queue`](http://nginx.org/en/docs/http/ngx_http_upstream_module.html#queue) for additional information.
```
Comment (Rulox) [inline docs/virtualserver-and-virtualserverroute.md]:
```suggestion
See [`queue`](http://nginx.org/en/docs/http/ngx_http_upstream_module.html#queue) directive for additional information.
```

File: internal/configs/virtualserver_test.go [hunk 4]
```diff
@@ -2095,3 +2094,94 @@ func TestCreateEndpointsFromUpstream(t *testing.T) {
...
+				Name: "test-upstream-queue-nil",
+			},
+			msg: "upstream queue with nil for OSS",
+		},
+	}
+
+	for _, test := range tests {
+		result := generateUpstream(test.name, test.upstream, false, []string{}, &ConfigParams{}, test.isPlus)
+		if !reflect.DeepEqual(result, test.expected) {
+			t.Errorf("generateUpstreamWithQueue() returned %v but expected %v for the case of %v", result, test.expected, test.msg)
```
Comment (Rulox) [inline internal/configs/virtualserver_test.go]:
```suggestion
			t.Errorf("generateUpstream() returned %v but expected %v for the case of %v", result, test.expected, test.msg)
```

File: pkg/apis/configuration/validation/validation.go [hunk 5]
```diff
@@ -815,3 +816,23 @@ func rejectPlusResourcesInOSS(upstream v1alpha1.Upstream, idxPath *field.Path, i
...
+	}
+
+	if !isPlus {
+		allErrs = append(allErrs, field.Forbidden(fieldPath, "queue is only supported in NGINX Plus"))
+		return allErrs
+	}
+
+	allErrs = append(allErrs, validateTime(queue.Timeout, fieldPath.Child("timeout"))...)
+	if queue.Size <= 0 {
+		allErrs = append(allErrs, field.Required(fieldPath.Child("size"), "must be positive and greater than 0"))
```
Comment (Rulox) [inline pkg/apis/configuration/validation/validation.go]:
```suggestion
		allErrs = append(allErrs, field.Required(fieldPath.Child("size"), "must be positive"))
```

I think `"must be positive and greater than 0"` sounds redundant, what about `"must be positive"` instead? a positive number is already greater than 0 so I don't think the second part adds any information. Also for consistency (eg `allErrs = append(allErrs, field.Invalid(fieldPath.Child("number"), buff.Number, "must be positive"))`

**Pull request merged**